File: modules/ballerina-core/src/main/java/org/ballerinalang/model/AnnotationAttachment.java
/*
 * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.ballerinalang.model;

import java.util.HashMap;
import java.util.Map;

/**
 * {@code Annotation} represents an Annotation node in Ballerina.
 * <p>
 * Annotation can be associated with various Ballerina concepts like Service, Resource, Functions, etc.
 *
 * @see <a href="https://github.com/wso2/ballerina/blob/master/docs/SyntaxSummary.md">Ballerina Syntax Summary</a>
 * @since 0.8.0
 */
public class AnnotationAttachment implements Node {

    private String name;
    private String pkgName;
    private String pkgPath;
    private SymbolName symbolName;
    private Map<SymbolName, AnnotationAttributeValue> attributeNameValPairs = new HashMap<>();
    private NodeLocation location;
    AttachmentPoint attachedPoint;
    
    public AnnotationAttachment(NodeLocation location, SymbolName name, String pkgName, String pkgPath,
            Map<SymbolName, AnnotationAttributeValue> fieldValPairs) {
        this.location = location;
        this.symbolName = name;
        this.pkgName = pkgName;
        this.pkgPath = pkgPath;
        this.attributeNameValPairs = fieldValPairs;
    }

    /**
     * Get name of the annotation.
     *
     * @return name of the annotation
     */
    public String getName() {
        return name;
    }

    public String getPkgName() {
        return pkgName;
    }

    public String getPkgPath() {
        return pkgPath;
    }
    
    /**
     * Get symbol name of the annotation.
     *
     * @return symbol name of the annotation
     */
    public SymbolName getSymbolName() {
        return this.symbolName;
    }

    /**
     * Get Key-Value pairs of fields in the annotation.
     *
     * @return all Key-Value pairs
     */
    public Map<SymbolName, AnnotationAttributeValue> getAttributeNameValuePairs() {
        return attributeNameValPairs;
    }

    /**
     * Get the value of the Key-Value pair.
     *
     * @param attributeName attribute name
     * @return value of the attribute
     */
    public AnnotationAttributeValue getValue(String attributeName) {
        return attributeNameValPairs.get(attributeName);
    }

    @Override
    public void accept(NodeVisitor visitor) {
        visitor.visit(this);
    }

    @Override
    public NodeLocation getNodeLocation() {
        return location;
    }
    
    /**
     * Set the construct where this annotation is attached.
     * 
     * @param attachedPoint 
     */
    public void setAttachedPoint(AttachmentPoint attachedPoint) {
        this.attachedPoint = attachedPoint;
    }
    
    /**
     * Get the construct where this annotation is attached.
     * 
     * @return the attachedPoint
     */
    public AttachmentPoint getAttachedPoint() {
        return attachedPoint;
    }

    /**
     * Builds an Annotation from parser events.
     *
     * @since 0.8.0
     */
    public static class AnnotationBuilder {
        private NodeLocation location;
        private SymbolName name;
        private String pkgName;
        private String pkgPath;
        private Map<SymbolName, AnnotationAttributeValue> attributeNameValPairs = new HashMap<>();

        /**
         * Set the source location of the annotation.
         * 
         * @param location Source location of the annotation
         */
        public void setNodeLocation(NodeLocation location) {
            this.location = location;
        }

        /**
         * Set the name of the annotation.
         * 
         * @param name Name of the annotation
         */
        public void setName(SymbolName name) {
            this.name = name;
        }

        /**
         * Set the package name of the annotation.
         * 
         * @param pkgName Package name of the annotation
         */
        public void setPkgName(String pkgName) {
            this.pkgName = pkgName;
        }

        /**
         * Set the package path of the annotation.
         * 
         * @param pkgPath Package path of the annotation
         */
        public void setPkgPath(String pkgPath) {
            this.pkgPath = pkgPath;
        }
        
        /**
         * Add a attribute name-value pair for the annotation.
         * 
         * @param name Name of the attribute
         * @param value Value of the attribute
         */
        public void addAttributeNameValuePair(SymbolName name, AnnotationAttributeValue value) {
            this.attributeNameValPairs.put(name, value);
        }

        public AnnotationAttachment build() {
            return new AnnotationAttachment(location, name, pkgName, pkgPath, attributeNameValPairs);
        }
    }
}

File: modules/web/js/ballerina/env/annotation-attachment.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import _ from 'lodash';
import EventChannel from 'event_channel';

/**
 * A class that represents an annotation attachment in ballerina-env.
 * 
 * @class AnnotationAttachment
 * @extends {EventChannel}
 */
class AnnotationAttachment extends EventChannel {
    /**
     * Creates an instance of Annotation.
     * @param {Object} args Object for creating an annotation in ballerina-env.
     * @param {string} args.packageName The name of the package of the annotation. This is the suffix. Example: http
     * @param {string} args.packagePath The full package name of the annotation. Example: ballerina.net.http
     * @param {string} args.name The name or the identifier of the annotation. Example: GET
     * @param {string} args.attachmentPoint The place which the annotation can be attached to. Example: "service", "resource".
     * 
     * @memberof AnnotationAttachment
     */
    constructor(args) {
        super(args);
        this._packageName = _.get(args, 'packageName');
        this._packagePath = _.get(args, 'packagePath');
        this._name = _.get(args, 'name');
        this._attachmentPoint = _.get(args, 'attachmentPoint');

        /**
         * Stores the attribute pair values of the annotation. new Map<string, AnnotationAttributeValue>().
         * @type {Map}
         */
        this._attributeNameValPairs = new Map();
    }

    setPackageName(packageName) {
        this._packageName = packageName;
    }

    getPackageName() {
        return this._packageName;
    }

    setPackagePath(packagePath) {
        this._packagePath = packagePath; 
    }

    getPackagePath() {
        return this._packagePath;
    }

    /**
    * sets the name
    * @param {string} name
    */
    setName(name) {
        var oldName = this._name;
        this._name = name;
        this.trigger('name-modified', name, oldName);
    }

    /**
     * returns the name
     * @returns {string}
     */
    getName() {
        return this._name;
    }

    setAttachmentPoint(attachmentPoint) {
        this._attachmentPoint = attachmentPoint;
    }
    
    getAttachmentPoint() {
        return this._attachmentPoint;
    }

    setAttributeNameValPairs(attributeNameValPairs) {
        this._attributeNameValPairs = attributeNameValPairs;
    }

    getAttributeNameValPairs() {
        return this._attributeNameValPairs;
    }

    addAttributeNameValPair(key, annotationAttributeValue) {
        this._attributeNameValPairs.set(key, annotationAttributeValue);
    }

    getAnnotationAttributeValue(key) {
        return this._attributeNameValPairs.get(key);
    }

    removeAttributeNameValPair(key) {
        this._attributeNameValPairs.delete(key);
    }

    /**
    * sets values from a json object
    * @param {Object} jsonNode
    */
    initFromJson(jsonNode) {
        var self = this;
        this.setName(jsonNode.name);

        _.each(jsonNode.attachmentPoints, function (attachmentPoint) {
            self.addAttachmentPoints(attachmentPoint);
        });

    }

}

export default AnnotationAttachment;